zircon.core/common/src/main/kotlin/org/hexworks/zircon/api/util/BSPTree.kt

```diff
@@ -0,0 +1,103 @@
+package org.hexworks.zircon.api.util
+
+import org.hexworks.zircon.api.data.Position
+import org.hexworks.zircon.api.data.Rect
+import org.hexworks.zircon.api.data.Size
+import kotlin.jvm.JvmStatic
+import kotlin.random.Random
+
+class BSPTree(var parent: BSPTree?, rec: Rect) {
+
+    companion object {
+        @JvmStatic
+        var minSize = 12
+
+        fun toMatrix(array: Array<CharArray>, BSPTrees: MutableList<BSPTree?>) {
+            var nbr = 48
+
+            for(BSPTree: BSPTree? in BSPTrees) {
+                if(BSPTree != null) {
+
+                    var char = nbr.toChar()
+                    var rec = BSPTree.room
+                    println(char + " " +BSPTree.boundingBox)
+                    if(rec != null) {
+
+                        for (y in rec.position.y..rec.position.y + rec.height) {
+                            for (x in rec.position.x..rec.position.x + rec.width) {
+
+                                array[y][x] = char
+                            }
+                        }
+                    }
+                }
+                nbr++
+            }
+        }
+
+        fun collectRooms(BSPTree: BSPTree?, list: MutableList<BSPTree?>) {
+            if(BSPTree?.leftBSPTree == null) {
+                list.add(BSPTree)
+            } else {
+                collectRooms(BSPTree.leftBSPTree, list)
```

Comment:
When you add `Maybe` and do away with `null`s you can use the `map` function to do something with the value when it is present like this:

```kotlin
BSPTree.leftBSPTree.map {
                    collectRooms(it, list)
                }
                BSPTree.rightBSPTree.map {
                    collectRooms(it, list)
                }
```